text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,875 @@
+  <div>Hello {{name}}!</div>
+</template>
+
+<script>
+export default {
+  data() {
+    return {
+      name: 'World'
+    }
+  }
+}
+</script>
+```
+
+### Module Imports in Templates
+
+One benefit of JSX is that it leverages JavaScript's existing scope
+system. React components are JavaScript values that can be imported and
+referenced like any other JavaScript value. If a binding would be in scope
+in JavaScript, it's in scope in JSX:
+
+```js
+import { Component } from 'react';
+import OtherComponent from './OtherComponent';
+
+export default class extends Component {
+  render() {
+    return (
+      <div>
+        // We know OtherComponent is in scope because it was imported at the top
+        // of the file, so we can invoke it here as a component.
+        <OtherComponent />
+      </div>
+    );
+  }
+}
+```
+
+This is a bit nicer than Vue's SFCs, where JavaScript code and template code
+happen to be in the same file but otherwise interact no differently than if
+they were still in separate files. Developers must learn a proprietary Vue
+API for explicitly copying values from JavaScript's scope into the template's
+scope, and component names can be different between where they are imported
+(`OtherComponent`) and where they are invoked (`other-component`):
+
+```html
+<template>
+  <div>
+    <other-component />
+  </div>
+</template>
+
+<script>
+import OtherComponent from './OtherComponent.vue'
+
+export default {
+  components: {
+    OtherComponent
+  }
+}
+</script>
+```
+
+Neither of these approaches is exactly right for Ember. Ideally, we'd find a
+way to "bend the curve" of tradeoffs: maintaining the performance benefits of
+templates, while gaining productivity and learnability by having those
+templates seamlessly participate in JavaScript's scoping rules.
+
+### Unlocking Experimentation
+
+Rather than deciding on the best syntax for single-file components and
+template imports upfront, this RFC proposes new low-level APIs that addons
+can use as compile targets to implement experimental file formats.
+
+For example, imagine a file format that uses a frontmatter-like syntax to
+combine template and JavaScript sections in a single file:
+
+```hbs
+---
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import titleize from './helpers/titleize';
+import BlogPost from './components/blog-post';
+
+export default class MyComponent extends Component {
+  @service session;
+}
+---
+
+{{#let this.session.currentUser as |user|}}
+  <BlogPost @title={{titleize @model.title}} @body={{@model.body}} @author={{user}} />
+{{/let}}
+```
+
+Or imagine this syntax, where a component's template is located syntactically
+within the class body:
+
+```js
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import titleize from './helpers/titleize';
+import BlogPost from './components/blog-post';
+
+export default class MyComponent extends Component {
```

Comment:
One of things that I really like about this example, and one of the things that I think/hope the SFC feature will enable is the ability to define multiple components within a single file. (I know, already consolidating from two files to one, and now I am asking to be able to consolidate from n files to one.)

Reply:
Yea, I want this world too. I'm actually a bit more excited about the possibility to have multiple helpers along side components. I think helpers are under utilized because the process of creating a new one is so heavy.